deku-derive/src/macros/deku_read.rs

```diff
@@ -262,38 +279,71 @@ fn emit_field_read(
             quote! { (#v) }
         })
         .or_else(|| Some(quote! { Result::<_, DekuError>::Ok }));
-    let field_count = &f.count;
     let field_ident = f.get_ident(i, true);
+    let hidden_field_ident = gen_hidden_field_ident(field_ident.clone());
 
     let field_read_func = if field_reader.is_some() {
         quote! { #field_reader }
-    } else if field_count.is_some() {
-        quote! { DekuRead::read(rest, input_is_le, field_bits, Some(usize::try_from(#field_count)?)) }
     } else {
-        quote! { DekuRead::read(rest, input_is_le, field_bits, None) }
+        let mut read_args = Vec::with_capacity(3);
+        if let Some(field_is_le) = field_is_le {
+            read_args.push(quote! {#field_is_le});
+        }
+        if let Some(field_bits) = f.bits {
+            read_args.push(quote! {#field_bits})
+        }
+        if let Some(ctx) = &f.ctx {
+            read_args.push(quote! {#ctx});
+        }
+
+        // Because `impl DekuRead<(bool, usize)>` but `impl DekuRead<bool>`(not tuple)
+        let read_args = if read_args.len() == 1 {
+            let arg = &read_args[0];
+            quote! {#arg}
+        } else {
+            quote! {#(#read_args),*}
+        };
+
+        // Count is special, we need to generate `(count, (other, ..))` for it.
+        if let Some(field_count) = &f.count {
+            // The count has same problem, when it isn't a copy type, the field will be moved.
+            // e.g. struct FooBar {
+            //   a: Baz // a type implement `Into<usize>` but not `Copy`.
+            //   #[deku(count = "a") <-- Oops, use of moved value: `a`
+            //   b: Vec<_>
+            // }
+            quote! {DekuRead::read(rest, (usize::try_from(#field_count << 0usize)?, (#read_args)))}
```

Comment:
Bit confused here, why is `#field_count << 0usize` needed?

Reply:
It's for type converting. Since `TryFrom` doesn't implement for &T(and `#field_cound` might be `123_i32` or `&i32`), I need to find a way to convert it to usize(`&i32 << usize == i32`). But I find here is a better way: using `usize::try_from(*(#field_cound.borrow()))`, I will fix it.